sdk/src/event.rs

```diff
@@ -1,12 +1,56 @@
 // Copyright 2021-2023 Protocol Labs
 // SPDX-License-Identifier: Apache-2.0, MIT
-use fvm_shared::event::ActorEvent;
-
 use crate::{sys, SyscallResult};
+use byteorder::{ByteOrder, LittleEndian};
+use fvm_shared::event::ActorEvent;
 
 pub fn emit_event(evt: &ActorEvent) -> SyscallResult<()> {
-    let encoded = fvm_ipld_encoding::to_vec(evt).expect("failed to marshal actor event");
-    let entries = encoded.as_slice();
+    // we manually serialize the ActorEvent (not using CBOR) into three byte arrays so
+    // we can accurately charge gas without needing to parse anything inside the FVM
+    const BYTES_PER_ENTRY: usize = 24;
+
+    let mut total_key_len: usize = 0;
+    let mut total_val_len: usize = 0;
+
+    let mut v = vec![0u8; evt.entries.len() * BYTES_PER_ENTRY];
+    for i in 0..evt.entries.len() {
+        let e = &evt.entries[i];
+        let offset = i * BYTES_PER_ENTRY;
+
+        let view = &mut v[offset..offset + BYTES_PER_ENTRY];
+        LittleEndian::write_u64(&mut view[..8], e.flags.bits());
+        LittleEndian::write_u64(&mut view[8..16], e.codec);
+        LittleEndian::write_u32(&mut view[16..20], e.key.len() as u32);
+        LittleEndian::write_u32(&mut view[20..24], e.value.len() as u32);
```

Comment:
We should be able to use transmutes here as well. As a matter of fact, we should be able to just:

1. Make an array of `EntryFixed`.
2. Transmute by-reference.

....

Actually, we _should_ be able to just take `*mut EventFixed` (pointer to an _array_) in the syscall ABI itself and let rust do the transmute for us. Welcome to C!